Rl cleanup

Rlamboll opened the pull request:
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Example added (either to an existing notebook or as a new notebook, where applicable)
- [x] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <https://github.com/znicholls/silicone/pull/XX>`_) Added feature which does something
- (`#XX <https://github.com/znicholls/silicone/pull/XX>`_) Fixed bug identified in (`#YY <https://github.com/znicholls/silicone/issues/YY>`_)
```

Rlamboll commented:
It's in steps, yes, as it will only return values found in the y dataset. You can see the answer to this question if you've ever run the plotting.py function (/script that runs this) on CO2 vs CO2 with quantiles on - it's pretty linear and as expected in the center of the distribution for all reasonable quantiles, but at the edges of the data there's less data so quantiles are pulled in the direction of the average value and are notably noisy.

znicholls commented:
> It's in steps, yes, as it will only return values found in the y dataset. You can see the answer to this question if you've ever run the plotting.py function (/script that runs this) on CO2 vs CO2 with quantiles on - it's pretty linear and as expected in the center of the distribution for all reasonable quantiles, but at the edges of the data there's less data so quantiles are pulled in the direction of the average value and are notably noisy.

Is it worth adding a test for this? Or putting in one of the example notebooks? The reason I ask is that if you use the infiller to infill a timeseries you already have, you won't get back what you put in which can be confusing so it would be good to document why.

Rlamboll commented:
If you have equally-spaced datapoints, match them up with the boxes and have very rapidly-decaying weights, the 0.5 quantile will be the original value. You can't guarantee it otherwise, no.

On 30 Oct 2019 00:43, Zeb Nicholls <notifications@github.com> wrote:

It's in steps, yes, as it will only return values found in the y dataset. You can see the answer to this question if you've ever run the plotting.py function (/script that runs this) on CO2 vs CO2 with quantiles on - it's pretty linear and as expected in the center of the distribution for all reasonable quantiles, but at the edges of the data there's less data so quantiles are pulled in the direction of the average value and are notably noisy.

Is it worth adding a test for this? Or putting in one of the example notebooks? The reason I ask is that if you use the infiller to infill a timeseries you already have, you won't get back what you put in which can be confusing so it would be good to document why.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fznicholls%2Fsilicone%2Fpull%2F20%3Femail_source%3Dnotifications%26email_token%3DACKUXI33223YIBOEBFIMUTLQRDKC7A5CNFSM4JGBLF3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECSROCA%23issuecomment-547690248&data=02%7C01%7C%7C348f57c5d39b46b6cce608d75cd22e45%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637079930089069194&sdata=9AfGYozBsXrL3r7rZ6eXqJrvS9tnH7%2B2I5Lv0TBpE54%3D&reserved=0>, or unsubscribe<https://eur04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACKUXIZ2XA57PHHFOKPYIRDQRDKC7ANCNFSM4JGBLF3A&data=02%7C01%7C%7C348f57c5d39b46b6cce608d75cd22e45%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637079930089079203&sdata=yjgzjZVknXQiex56QCKj4%2FNCvYHjXU7Dp%2Bnr%2BCbc7y4%3D&reserved=0>.

znicholls commented:
Ok I think it would be good to explain that in a notebook then. We can either do that in this PR or we can open an issue and then do it in a later PR, I'm happy either way (and conscious of time).

Rlamboll commented:
OK, I've added stuff to the notebook to demonstrate the quirks of the windows value and retooled a test to check that you get out what you put in when there's nothing to average over. I've also removed the warning message.

znicholls commented:
@Rlamboll I also didn't understand the last part of the notebook. Can you please add some more explanatory plots (or we make an issue and deal with it later).

znicholls commented:
@Rlamboll once we've dealt with #24 and the review comments let's merge above. We can leave the notebook for #25

znicholls commented on src/silicone/database_crunchers/quantile_rolling_windows.py:
```diff
@@ -164,46 +158,51 @@ def derive_relationship(
+                    db_time
+                ] = lambda unused_variable, ys=ys, cumsum_weights=cumsum_weights, quantile=quantile: min(
+                    ys[cumsum_weights >= quantile]
+                )
```
Can we change the lines above to the below (which is easier to understand I think)

```python
                def same_x_val_workaround(_, ys=ys, cumsum_weights=cumsum_weights, quantile=quantile):
                    return min(ys[cumsum_weights >= quantile])

                derived_relationships[
                    db_time
                ] = same_x_val_workaround
```

Rlamboll replied:
This is in one of the new commits

Rlamboll commented on src/silicone/database_crunchers/quantile_rolling_windows.py:
```diff
@@ -164,44 +160,49 @@ def derive_relationship(
+                derived_relationships[db_time] = scipy.interpolate.interp1d(
+                    db_time_table.columns.values.squeeze(),
+                    db_time_table.loc[(db_time, quantile), :].values.squeeze(),
+                    bounds_error=False,
```
On balance I think it's probably more useful if we just use the extremal values, I'll put that in

znicholls replied:
beauty

znicholls commented on tests/integration/crunchers/test_cruncher_quantile_rolling_windows.py:
```diff
@@ -167,6 +167,24 @@ def test_numerical_relationship(self):
+        ind = modify_extreme_db["value"].idxmax
+        modify_extreme_db["value"].loc[ind] += 10
+        extreme_crunched = res(modify_extreme_db)
+        assert all(crunched["value"] == extreme_crunched["value"])
```
can you put a comment above this explaining? Something like, "filling the df with the extreme value should give the same result as filling the df without the extreme value as our filler's extrapolation is constant"
